### PR TITLE
Refactor of ecfNode logic and updated documentation

### PR DESCRIPTION
**Pull request by KyleNevins-NOAA**
**Description**
This PR was designed around a redesign of the workflow_generator introduced in the last merge. The last one had the essential capabilities but relied too much on if/else loops as opposed to putting more into the object oriented class approach. It also limited the ability for families to be a loop. This PR reworks the classes so more is offloaded from the individual if/else statements in the methods and more is handled by the class objects which makes the code much cleaner. It also allows family objects to be loops as well as then passing down that context to the child task or triggers. 
**Type of change**

Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**
The setup was run against both the prod.yml by running 
`python3 setup_workflow.py --expdir ../parm/config/ --ecflow-config prod.yml`
and then inserting that into a local running copy of the ecflow application running on my desktop with 
`ecflow_client --port 3142 --load=ecflow_suite.def`
The items successfully loaded into the running ecflow server. It has NOT been run on the operational or development environments against actual scripts since the current script structure was changing. 
I also used the ecflow_build.yml file to test as well, similarly with 
`python3 setup_workflow.py --expdir ../parm/config/` since the ecflow_build.yml file is the default and then loading it into the locally runnng ecflow server. Then loaded it up into the ecflow server with: 
`ecflow_client --port 3142 --load=ecflow_suite.def`

- [X] Test the prod.yml configuration file
- [X] Test the ecflow_build.yml configuration file
- [X] Load test ecflow_suite.def file into a locally running copy of the ecflow server. 
  
**Checklist**
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- **N/A** - New and existing tests pass with my changes
- **N/A** -Any dependent changes have been merged and published

### CODE REVIEW

File: workflow_generator/ecflow_setup/ecflow_definitions.py [hunk 1]
```diff
@@ -741,140 +781,104 @@ def add_task_triggers(self, task, triggers, suite_array):
...
-                        sys.exit(1)
-                else:
-                    process_trigger(triggerTaskNode.get_name(), triggerTaskNode, task)
+                            self.add_trigger(trigger_name, node_name,
+                                             suite=suite,
+                                             suite_array=suite_array,
+                                             operand=operand)
 
 
 class ecfNode():
```
Comment (WalterKolczynski-NOAA) [inline workflow_generator/ecflow_setup/ecflow_definitions.py]: All these class names should be camel case and start with a capital letter. This is something I missed last time. So:

EcflowSuite (or maybe ECFlowSuite)
EcfNode
EcFamilyNode

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Looks fine. Only thing I noticed was the class names aren't correct, but that's a minor thing.

File: workflow_generator/ecflow_setup/ecflow_definitions.py (hunk 1 above)
Comment (KyleNevins-NOAA) [reply workflow_generator/ecflow_setup/ecflow_definitions.py]: I forgot to respond to this that I had updated the camel case appropriately with a follow on commit.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
Is this ready to merge, or did you want to add something more to the documentation first?

**Comment by KyleNevins-NOAA** [comment]
I was comfortable merging. I went over the guides and documentation again yesterday and didn't see any issues and the guides made sense to me. I hope they make sense to you all as well. From my perspective it is ready to merge and use. Other than the access to stmp, I was ready to run experiements with the definition files created by it, using Lin's repository as the script repo so I didn't need to modify my scripts. I also created dummy jobs and scripts and was able to successfully create a yaml, use a custom script repo path and execute an ecflow suite so it all works together as it stands now. 

>

**Review by aerorahul — APPROVED** [review]
looks good.
@ryanlong1004 will be extending this work.

**Pull request merged**